src/Server/KeeperTCPHandler.cpp

```diff
@@ -290,10 +293,23 @@ Poco::Timespan KeeperTCPHandler::receiveHandshake(int32_t handshake_length, bool
 
     Coordination::read(protocol_version, *in);
 
-    if (protocol_version != Coordination::ZOOKEEPER_PROTOCOL_VERSION && protocol_version != Coordination::ZOOKEEPER_PROTOCOL_VERSION_WITH_COMPRESSION)
+    if (protocol_version != Coordination::ZOOKEEPER_PROTOCOL_VERSION
+        && protocol_version < Coordination::ZOOKEEPER_PROTOCOL_VERSION_WITH_COMPRESSION
+        && protocol_version > Coordination::ZOOKEEPER_PROTOCOL_VERSION_WITH_XID_64)
+    {
         throw Exception(ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT, "Unexpected protocol version: {}", toString(protocol_version));
+    }
 
-    use_compression = (protocol_version == Coordination::ZOOKEEPER_PROTOCOL_VERSION_WITH_COMPRESSION);
+    if (protocol_version == Coordination::ZOOKEEPER_PROTOCOL_VERSION_WITH_COMPRESSION)
+    {
+        use_compression = true;
+    }
+    else if (protocol_version >= Coordination::ZOOKEEPER_PROTOCOL_VERSION_WITH_XID_64)
+    {
+        close_xid = Coordination::CLOSE_XID_64;
+        use_xid_64 = true;
```

Comment:
Why `true` is set for new protocol version unconditionally without taking into account config option? I'd expect setting it from config for new protocol and explicit disabling for old one.

_UPD_: seems in that case we will *try* to find second part for xid, but if it's not present will use 32 bit variant depending on what other side had sent, is it correct ?

Reply:
So client defines if XID will be 32 or 64 bit.
Based on the config, it will send a different protocol version (if `use_xid_64` is set to `true`, `ZOOKEEPER_PROTOCOL_VERSION_WITH_XID_64` will be used).
Why unconditionally? Because I see no reason to use 32bit XID if Keeper supports 64bit.

> UPD: seems in that case we will try to find second part for xid, but if it's not present will use 32 bit variant depending on what other side had sent, is it correct ?

This is only true for serializing and deserializing RAFT log entries, but for communication the XID will be sent in a chunk as 32 or 64 bit integer, depending on the protocol version used.